flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java

```diff
@@ -0,0 +1,697 @@
+        List<StreamNode> enterPoints = new ArrayList<>();
+        for (StreamNode sourceNode : sourceNodes) {
+            if (isSourceChainable(sourceNode, streamGraph)) {
+                StreamEdge outEdge = sourceNode.getOutEdges().get(0);
+                StreamNode startNode = streamGraph.getStreamNode(outEdge.getTargetId());
+                chainedStreamNodesMap
+                        .computeIfAbsent(startNode, k -> new ArrayList<>())
+                        .add(sourceNode);
+                enterPoints.add(startNode);
+            } else {
+                chainedStreamNodesMap.computeIfAbsent(sourceNode, k -> new ArrayList<>());
+                enterPoints.add(sourceNode);
+            }
+        }
+        return enterPoints;
+    }
+
+    private void traverseFullGraph(
+            Integer startNodeId,
+            Integer currentNodeId,
+            Map<Integer, List<StreamEdge>> transitiveNonChainableOutEdgesMap,
+            Map<StreamNode, List<StreamNode>> topologicallySortedChainedStreamNodesMap,
+            Set<Integer> visitedStartNodes) {
+        if (visitedStartNodes.contains(startNodeId)) {
+            return;
+        }
+        List<StreamEdge> chainableOutputs = new ArrayList<>();
+        List<StreamEdge> nonChainableOutputs = new ArrayList<>();
+        StreamNode currentNode = streamGraph.getStreamNode(currentNodeId);
+        StreamNode startNode = streamGraph.getStreamNode(startNodeId);
+        for (StreamEdge streamEdge : currentNode.getOutEdges()) {
+            if (isChainable(streamEdge, streamGraph)) {
+                chainableOutputs.add(streamEdge);
+            } else {
+                nonChainableOutputs.add(streamEdge);
+                topologicallySortedChainedStreamNodesMap.computeIfAbsent(
+                        streamGraph.getStreamNode(streamEdge.getTargetId()),
+                        k -> new ArrayList<>());
+            }
+        }
+
+        transitiveNonChainableOutEdgesMap
+                .computeIfAbsent(startNodeId, k -> new ArrayList<>())
+                .addAll(nonChainableOutputs);
+
+        topologicallySortedChainedStreamNodesMap.get(startNode).add(currentNode);
+
+        for (StreamEdge chainable : chainableOutputs) {
+            traverseFullGraph(
+                    startNodeId,
+                    chainable.getTargetId(),
+                    transitiveNonChainableOutEdgesMap,
+                    topologicallySortedChainedStreamNodesMap,
+                    visitedStartNodes);
+        }
+        for (StreamEdge nonChainable : nonChainableOutputs) {
+            traverseFullGraph(
+                    nonChainable.getTargetId(),
+                    nonChainable.getTargetId(),
+                    transitiveNonChainableOutEdgesMap,
+                    topologicallySortedChainedStreamNodesMap,
+                    visitedStartNodes);
+        }
+        if (currentNodeId.equals(startNodeId)) {
```

Comment:
Why only skipping start nodes? And why adding it to the visited nodes in the end? Looks to me this may result in multiple visitations to one node.